syntax = "proto3";

package search;

import "google/protobuf/empty.proto";

service Index {
    rpc takeNext (google.protobuf.Empty) returns (TakeNextResponse);
    rpc putNew (PutNewRequest) returns (google.protobuf.Empty);
    rpc addToIndexWords (AddToIndexRequestWords) returns (google.protobuf.Empty);
    rpc addToIndexLinks (AddToIndexRequestLinks) returns (google.protobuf.Empty);
    rpc searchWord (SearchWordRequest) returns (SearchWordResponse);
    rpc getFullIndex (google.protobuf.Empty) returns (FullIndexResponse);
    rpc registerIndexBarrel (IndexBarrelRequest) returns (ValidRegister);
    rpc getIndexBarrels (google.protobuf.Empty) returns (IndexBarrelInfo);
}

message PutNewRequest {
    string url = 1;
}

message AddToIndexRequestWords {
    string word = 1;
    string url = 2;
    bool from_multicast = 3;
}

message AddToIndexRequestLinks {
    string url = 1;
    string link = 2;
    bool from_multicast = 3;
}

message SearchWordRequest {
    string words = 1;
}

message SearchWordResponse {
    repeated string urls = 1;
}

message TakeNextResponse {
    string url = 1;
}

message FullIndexResponse {
    repeated IndexEntry palavras = 1;
    repeated IndexEntry urls = 2;
}

message IndexEntry {
    string param0 = 1;
    string param1 = 2;
}

message IndexBarrelRequest {
    string host = 1;
    string port = 2;
}

message ValidRegister{
    bool valid = 1;
}

message IndexBarrelInfo{
    repeated string indexInfos = 1;
}

